nearcore/src/state_sync.rs

```diff
@@ -773,27 +773,28 @@ impl StateDumper {
             .iter()
             .map(|(shard_id, _)| (uploader.clone(), *shard_id))
             .collect::<Vec<_>>();
-        let headers_stored = tokio_stream::iter(shards)
-            .filter_map(|(uploader, shard_id)| async move {
-                let stored = uploader.header_stored(shard_id).await;
-                if stored { Some(futures::future::ready(shard_id)) } else { None }
-            })
-            .buffer_unordered(10)
-            .collect::<Vec<_>>()
-            .await;
-        for shard_id in headers_stored {
-            tracing::info!(
-                target: "state_sync_dump", %shard_id, epoch_height = %dump.epoch_height,
-                "Header already saved to external storage."
-            );
-            let s = dump.dump_state.get_mut(&shard_id).unwrap();
-            s.header_to_dump = None;
+        async move {
+            let headers_stored = tokio_stream::iter(shards)
+                .filter_map(|(uploader, shard_id)| async move {
+                    let stored = uploader.header_stored(shard_id).await;
+                    if stored { Some(futures::future::ready(shard_id)) } else { None }
+                })
+                .buffer_unordered(10)
+                .collect::<Vec<_>>()
+                .await;
+            for shard_id in headers_stored {
+                tracing::info!(
+                    target: "state_sync_dump", %shard_id, epoch_height = %dump.epoch_height,
+                    "Header already saved to external storage."
+                );
+                let s = dump.dump_state.get_mut(&shard_id).unwrap();
+                s.header_to_dump = None;
+            }
```

Comment:
Pre-existing but this could avoid collecting a vector unnecessarily by using a pattern along the lines of:

```suggestion
            let mut headers_stored = tokio_stream::iter(shards)
                .filter_map(|(uploader, shard_id)| async move {
                    let stored = uploader.header_stored(shard_id).await;
                    if stored { Some(futures::future::ready(shard_id)) } else { None }
                })
                .buffer_unordered(10);
            while let Some(shard_id) = headers_stored.next().await {
                tracing::info!(
                    target: "state_sync_dump", %shard_id, epoch_height = %dump.epoch_height,
                    "Header already saved to external storage."
                );
                let s = dump.dump_state.get_mut(&shard_id).unwrap();
                s.header_to_dump = None;
            }
```